Implement Tours Repository

Koshake opened the pull request:
Repository and data model for tours

Commit:
feat: model for tours data added
feat: repository for tours added
feat: methods for getting tours from server added

RomchaDev commented on model/src/main/java/com/demal/model/data/entity/tours/Tour.kt:
```diff
@@ -0,0 +1,17 @@
+    @field:SerializedName("country") val country: String,
+    @field:SerializedName("travelersCount") val travelersCount: Int,
+    @field:SerializedName("startDate") val startDate: String
+) : AppStateEntity
```
Зачем всюду добавлять SerializedName?

Koshake replied:
Можно спокойно убрать, не особо это нужно

RomchaDev replied:
Убери, если не нужно

Koshake replied:
fixed

RomchaDev commented on model/src/main/java/com/demal/model/data/entity/tours/Tours.kt:
```diff
@@ -0,0 +1,5 @@
+
+import com.demal.model.data.entity.AppStateEntity
+
+data class Tours(val toursList : List<Tour>) : AppStateEntity
```
Зачем этот класс?

Koshake replied:
Как обертка над List<Tour>, ведь он не является AppStateEntity и я потом не смогу его передать в HomeViewModel. 
Сейчас вижу, что я в итоге нигде не использую класс Tours, а надо бы из функции getTours в ToursRepository возвращать тогда не List<Tour>, а Tours

RomchaDev replied:
Ты уже это сделала?

Koshake replied:
fixed

RomchaDev commented on repository/src/main/java/com/demal/repository/repository/ToursRepository.kt:
```diff
@@ -0,0 +1,8 @@
+
+interface ToursRepository {
+    suspend fun getTours(sortBy: String, order: String) : List<Tour>
+    suspend fun getTourById(id: Int) : Tour
```
Думаю, стоит создать enum, для того, чтобы было понятно какие могут быть варианты для sortBy и для order. Я не уверен, где должно происходить преобразование из enum в string (Repository / DataSource). Погугли, то, что нагуглишь - мне брось, тоже почитаю.

Koshake replied:
Нашла такую интересную статью
https://medium.com/tedpark-developer/how-to-use-enum-with-retrofit-query-path-d9c14b93d68f

RomchaDev replied:
Отлично, надо реализовать

Koshake replied:
fixed

RomchaDev reviewed (changes requested):
Ответь на вопросы и сделай enum

Commit:
Merge remote-tracking branch 'origin/dev' into feature/tour-repository

# Conflicts:
#	app/src/main/java/com/demal/koin/modules/RepositoryModules.kt

Commit:
fix: image, category added to tours data model
feat: different data models for data from server and app data
feat: fiels "isFavorite" added
feat: mappers added
feat: enums for sorting and ordering added
feat: enum converter factory added to Retrofit

RomchaDev commented on model/src/main/java/com/demal/model/data/entity/tours/network/NetworkTour.kt:
```diff
@@ -0,0 +1,6 @@
+package com.demal.model.data.entity.tours.network
+
+data class NetworkTour(
+    val tour: TourResponse,
```
Наверное этот класс лучше назвать TourResponse, а тот, который сейчас так называетсся, назвать NetworkTour.

Koshake replied:
fixed

RomchaDev commented on model/src/main/java/com/demal/model/data/entity/tours/network/TourResponse.kt:
```diff
@@ -0,0 +1,18 @@
+package com.demal.model.data.entity.tours.network
+
+data class TourResponse(
+    val id: Int,
```
Уже писал по поводу названий

Koshake replied:
fixed

RomchaDev commented on repository/src/main/java/com/demal/repository/types/EnumConverterFactory.kt:
```diff
@@ -0,0 +1,26 @@
+                try {
+                    enum.javaClass.getField(enum.name)
+                        .getAnnotation(SerializedName::class.java)?.value
+                } catch (exception: Exception) {
```
Насколько я понимаю, если у поля нету аннотации SerializedName, то этот метод вернёт null. Если это действительно так (я могу ошибаться), то это нужно исправить. Надо, чтобы если у поля небыло SerialisedName, этот метод возвращал название поля.

Koshake replied:
да, точно! ты прав. Мне тогда надо либо добавить SerializedName к каждому полю enum, либо переименовать поля c заглавных в строчные буквы. Как лучше сделать?

RomchaDev replied:
Лучше добавить SerializedName. А если есть интерес и желание, то можно добавить в EnumConverterFactory поле namingPolicy и если туда передаётся значение CAMEL_CASE, то нужно преобразовать имя поля енама в camelCase. Но это совсем не обязательно.

Koshake replied:
fixed

RomchaDev reviewed (changes requested):
Ещё немного вопросов. Ты проверяла работоспособность этого репозитория? Если нет, то нужно проверить

Commit:
refactor: swap names TourResponse - NetworkTour
refactor: nullable types for tour data fields
refactor: domain data model removed
refactor: mappers removed
fix: SerializedName added to enums
fix: Enum name returned from EnumConverterFactory

RomchaDev commented on model/src/main/java/com/demal/model/data/entity/tours/network/NetworkTour.kt:
```diff
@@ -1,6 +1,18 @@
 package com.demal.model.data.entity.tours.network
 
 data class NetworkTour(
-    val tour: TourResponse,
```
Он уже не "Network", нужно переименовать. Для юзера я вообще не создавал отдельного пакета network, что, возможно, не хорошо. Но в network должны лежать только классы request и response. Tour нужно убрать из этого пакета.

Koshake replied:
Сама сомневаюсь как нейминг сделать. Теперь переименовать NetworkTour в TourResponse, а TourResponse в просто Tour?

RomchaDev replied:
NetworkTour в Tour

Koshake replied:
fixed

RomchaDev reviewed (changes requested):
Один небольшой вопрос по структуре папок

Commit:
refactor: rename NetworkTour - Tour

Commit:
refactor: rename TourResponse - Tour
refactor: move Tour from network

RomchaDev committed:
feat: changed working with favorites logic

RomchaDev committed:
refactor: deleted useless class

RomchaDev committed:
feat: added delete and post wish requests

RomchaDev committed:
refactor: fixed an error

RomchaDev committed:
refactor: implemented adding and deleting tours from local list

RomchaDev commented on repository/src/main/java/com/demal/repository/repository/UserRepositoryImpl.kt:
```diff
@@ -1,18 +1,33 @@
 class UserRepositoryImpl(
     private val remoteDataSource: RemoteDataSource,
-    private val tokenRepository : TokenRepository
+    private val responseRepository: UserRepositoryLocal,
```
Rename to userRepositoryLocal

RomchaDev commented on model/src/main/java/com/demal/model/data/entity/tours/ITour.kt:
```diff
@@ -0,0 +1,21 @@
+import com.demal.model.data.entity.tours.network.ImageResponse
+
+interface ITour {
+    val id: Int
```
Make some fields not nullable

RomchaDev reviewed (changes requested):
Resolve conversations

RomchaDev committed:
refactor: resolved some conversations
